File: src/bin/generate-mdn-data.rs
use std::{
    collections::{BTreeMap, BTreeSet},
    fs,
    io::Read,
    path::{Path, PathBuf},
    str::FromStr,
};

use anyhow::{Context, Result};
use clap::Parser;
use serde_json::Value;

const DEFAULT_BCD_VERSION: &str = include_str!("../../data/mdn-bcd.version");
const DEFAULT_OUTPUT_DIR: &str = "data/mdn";

#[derive(Debug, Parser)]
#[command(name = "generate-mdn-data")]
#[command(about = "Generate runtime RON data from MDN browser-compat-data")]
struct Args {
    /// Read MDN BCD data.json from disk instead of downloading it.
    #[arg(long)]
    input: Option<PathBuf>,

    /// MDN @mdn/browser-compat-data version to download.
    #[arg(long)]
    bcd_version: Option<String>,

    /// Directory where runtime RON files are written.
    #[arg(long, default_value = DEFAULT_OUTPUT_DIR)]
    output_dir: PathBuf,

    /// Runtime ids to generate: nodejs, deno, bun, safari, chrome, firefox.
    #[arg(
        long,
        value_delimiter = ',',
        default_value = "nodejs,deno,bun,safari,chrome,firefox"
    )]
    runtimes: Vec<String>,
}

fn main() -> Result<()> {
    let args = Args::parse();
    run(args)
}

fn run(args: Args) -> Result<()> {
    let bcd_version = args
        .bcd_version
        .unwrap_or_else(|| DEFAULT_BCD_VERSION.trim().to_string());
    let input = load_bcd_json(args.input.as_deref(), &bcd_version)?;
    let bcd: Value = serde_json::from_str(&input).context("failed to parse MDN BCD data.json")?;

    if let Some(actual) = bcd
        .pointer("/__meta/version")
        .and_then(Value::as_str)
        .filter(|actual| *actual != bcd_version)
    {
        anyhow::bail!("expected BCD {bcd_version}, got {actual}");
    }

    fs::create_dir_all(&args.output_dir)
        .with_context(|| format!("failed to create {}", args.output_dir.display()))?;

    for runtime_id in args.runtimes {
        let target = RuntimeTarget::from_bcd_id(&runtime_id)?;
        let features = extract_features(&bcd, target.bcd_id)?;
        let output =
            render_runtime_ron(target.runtime_name, target.bcd_id, &bcd_version, &features);
        let path = args.output_dir.join(format!("{}.ron", target.runtime_name));
        fs::write(&path, output).with_context(|| format!("failed to write {}", path.display()))?;
        println!(
            "generated {} features for {} from MDN BCD {} -> {}",
            features.len(),
            target.runtime_name,
            bcd_version,
            path.display()
        );
    }

    Ok(())
}

fn load_bcd_json(input: Option<&Path>, version: &str) -> Result<String> {
    if let Some(input) = input {
        return fs::read_to_string(input)
            .with_context(|| format!("failed to read {}", input.display()));
    }

    let url = format!("https://unpkg.com/@mdn/browser-compat-data@{version}/data.json");
    let response = ureq::get(&url)
        .call()
        .with_context(|| format!("failed to download {url}"))?;
    let mut text = String::new();
    response
        .into_reader()
        .read_to_string(&mut text)
        .context("failed to read downloaded MDN BCD data.json")?;
    Ok(text)
}

#[derive(Debug, Clone, Copy)]
struct RuntimeTarget {
    bcd_id: &'static str,
    runtime_name: &'static str,
}

impl RuntimeTarget {
    fn from_bcd_id(id: &str) -> Result<Self> {
        match id {
            "nodejs" | "node" => Ok(Self {
                bcd_id: "nodejs",
                runtime_name: "node",
            }),
            "deno" => Ok(Self {
                bcd_id: "deno",
                runtime_name: "deno",
            }),
            "bun" => Ok(Self {
                bcd_id: "bun",
                runtime_name: "bun",
            }),
            "safari" => Ok(Self {
                bcd_id: "safari",
                runtime_name: "safari",
            }),
            "chrome" => Ok(Self {
                bcd_id: "chrome",
                runtime_name: "chrome",
            }),
            "firefox" => Ok(Self {
                bcd_id: "firefox",
                runtime_name: "firefox",
            }),
            other => anyhow::bail!("unsupported BCD runtime id `{other}`"),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
struct GeneratedFeature {
    name: String,
    version: Version,
    detect: Vec<GeneratedDetectRule>,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
struct Version {
    major: u64,
    minor: u64,
    patch: u64,
}

impl FromStr for Version {
    type Err = anyhow::Error;

    fn from_str(value: &str) -> Result<Self> {
        let value = value.trim().trim_start_matches('v');
        if !value
            .chars()
            .next()
            .is_some_and(|first| first.is_ascii_digit())
        {
            anyhow::bail!("version is not numeric: {value}");
        }

        let mut parts = value.split('.');
        let major = parse_version_part(parts.next().unwrap_or("0"), value, "major")?;
        let minor = parse_version_part(parts.next().unwrap_or("0"), value, "minor")?;
        let patch = parse_version_part(parts.next().unwrap_or("0"), value, "patch")?;
        Ok(Self {
            major,
            minor,
            patch,
        })
    }
}

impl std::fmt::Display for Version {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(formatter, "{}.{}.{}", self.major, self.minor, self.patch)
    }
}

fn parse_version_part(part: &str, full: &str, label: &str) -> Result<u64> {
    if part.is_empty() || !part.chars().all(|ch| ch.is_ascii_digit()) {
        anyhow::bail!("invalid {label} version in `{full}`");
    }
    part.parse()
        .with_context(|| format!("invalid {label} version in `{full}`"))
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
enum GeneratedDetectRule {
    Global(String),
    MemberChain(String),
    Property(String),
}

fn extract_features(bcd: &Value, runtime_id: &str) -> Result<Vec<GeneratedFeature>> {
    let mut features = BTreeMap::<String, GeneratedFeature>::new();

    if let Some(api) = bcd.get("api") {
        walk_compat_tree(
            api,
            &mut Vec::new(),
            SourceKind::Api,
            runtime_id,
            &mut features,
        );
    }

    if let Some(builtins) = bcd.pointer("/javascript/builtins") {
        walk_compat_tree(
            builtins,
            &mut Vec::new(),
            SourceKind::JavascriptBuiltin,
            runtime_id,
            &mut features,
        );
    }

    Ok(features.into_values().collect())
}

#[derive(Debug, Clone, Copy)]
enum SourceKind {
    Api,
    JavascriptBuiltin,
}

fn walk_compat_tree(
    value: &Value,
    path: &mut Vec<String>,
    source: SourceKind,
    runtime_id: &str,
    features: &mut BTreeMap<String, GeneratedFeature>,
) {
    let Some(object) = value.as_object() else {
        return;
    };

    if let Some(compat) = object.get("__compat")
        && let Some(version) = support_version(compat, runtime_id)
    {
        let name = feature_name(path);
        if !name.is_empty() && is_runtime_surface_path(path) {
            let detect = detect_rules(source, path);
            if !detect.is_empty() {
                upsert_feature(
                    features,
                    GeneratedFeature {
                        name,
                        version,
                        detect,
                    },
                );
            }
        }
    }

    for (key, child) in object {
        if key == "__compat" {
            continue;
        }
        path.push(key.clone());
        walk_compat_tree(child, path, source, runtime_id, features);
        path.pop();
    }
}

fn upsert_feature(features: &mut BTreeMap<String, GeneratedFeature>, feature: GeneratedFeature) {
    match features.get(&feature.name) {
        Some(existing) if existing.version <= feature.version => {}
        _ => {
            features.insert(feature.name.clone(), feature);
        }
    }
}

fn support_version(compat: &Value, runtime_id: &str) -> Option<Version> {
    let support = compat.get("support")?.get(runtime_id)?;
    let statements: Vec<&Value> = match support {
        Value::Array(values) => values.iter().collect(),
        value => vec![value],
    };

    statements
        .into_iter()
        .filter(|statement| !has_runtime_flags(statement))
        .filter(|statement| !has_version_removed(statement))
        .filter_map(|statement| statement.get("version_added")?.as_str())
        .filter_map(|version| version.parse::<Version>().ok())
        .min()
}

fn has_runtime_flags(statement: &Value) -> bool {
    statement
        .get("flags")
        .and_then(Value::as_array)
        .is_some_and(|flags| !flags.is_empty())
}

fn has_version_removed(statement: &Value) -> bool {
    statement
        .get("version_removed")
        .is_some_and(|version| !version.is_null())
}

fn feature_name(path: &[String]) -> String {
    path.join(".")
}

fn is_runtime_surface_path(path: &[String]) -> bool {
    if path.is_empty() {
        return false;
    }

    if path.len() > 1 && path.first() == path.last() {
        return false;
    }

    path.iter()
        .all(|segment| is_runtime_surface_segment(segment))
}

fn is_runtime_surface_segment(segment: &str) -> bool {
    if segment.is_empty() || segment.starts_with("@@") || segment.contains('-') {
        return false;
    }

    if segment.contains('_') {
        return segment
            .chars()
            .all(|ch| ch == '_' || ch.is_ascii_digit() || ch.is_ascii_uppercase());
    }

    segment
        .chars()
        .all(|ch| ch == '$' || ch.is_ascii_alphanumeric())
}

fn detect_rules(source: SourceKind, path: &[String]) -> Vec<GeneratedDetectRule> {
    if path.is_empty() {
        return Vec::new();
    }

    let name = feature_name(path);
    let mut rules = BTreeSet::new();

    if path.len() == 1 {
        rules.insert(GeneratedDetectRule::Global(name.clone()));
    }

    rules.insert(GeneratedDetectRule::MemberChain(name));

    if matches!(source, SourceKind::JavascriptBuiltin)
        && path.len() >= 2
        && let Some(property) = path
            .last()
            .filter(|property| is_detectable_property(property))
    {
        rules.insert(GeneratedDetectRule::Property(property.clone()));
    }

    rules.into_iter().collect()
}

fn is_detectable_property(property: &str) -> bool {
    !property.starts_with("@@")
        && !property.starts_with("__")
        && property
            .chars()
            .next()
            .is_some_and(|first| first == '_' || first.is_ascii_alphabetic())
}

fn render_runtime_ron(
    runtime_name: &str,
    bcd_id: &str,
    bcd_version: &str,
    features: &[GeneratedFeature],
) -> String {
    let mut output = String::new();
    output.push_str(&format!(
        "//! Generated from MDN @mdn/browser-compat-data {bcd_version} runtime `{bcd_id}`.\n"
    ));
    output.push_str(
        "//! Regenerate with: cargo run --bin generate-mdn-data -- --output-dir data/mdn\n",
    );
    output.push_str("(\n");
    output.push_str("    schema: 1,\n");
    output.push_str(&format!("    runtime: {},\n", quote(runtime_name)));
    output.push_str("    features: [\n");

    for feature in features {
        output.push_str("        (name: ");
        output.push_str(&quote(&feature.name));
        output.push_str(", version: ");
        output.push_str(&quote(&feature.version.to_string()));
        output.push_str(", detect: [");
        for (index, rule) in feature.detect.iter().enumerate() {
            if index > 0 {
                output.push_str(", ");
            }
            output.push_str(&render_detect_rule(rule));
        }
        output.push_str("]),\n");
    }

    output.push_str("    ],\n");
    output.push_str(")\n");
    output
}

fn render_detect_rule(rule: &GeneratedDetectRule) -> String {
    match rule {
        GeneratedDetectRule::Global(name) => format!("Global({})", quote(name)),
        GeneratedDetectRule::MemberChain(name) => format!("MemberChain({})", quote(name)),
        GeneratedDetectRule::Property(name) => format!("Property({})", quote(name)),
    }
}

fn quote(value: &str) -> String {
    serde_json::to_string(value).expect("string escaping cannot fail")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn extracts_runtime_versions_from_bcd_shape() {
        let bcd = serde_json::json!({
            "__meta": { "version": "fixture" },
            "api": {
                "fetch": {
                    "__compat": {
                        "support": {
                            "nodejs": { "version_added": "18.0.0" },
                            "deno": { "version_added": "1.0" }
                        }
                    }
                }
            },
            "javascript": {
                "builtins": {
                    "Array": {
                        "toSorted": {
                            "__compat": {
                                "support": {
                                    "nodejs": { "version_added": "20.0.0" }
                                }
                            }
                        },
                        "options_parameter": {
                            "__compat": {
                                "support": {
                                    "nodejs": { "version_added": "21.0.0" }
                                }
                            }
                        },
                        "flagged": {
                            "__compat": {
                                "support": {
                                    "nodejs": {
                                        "version_added": "21.0.0",
                                        "flags": [{ "name": "flag" }]
                                    }
                                }
                            }
                        }
                    },
                    "Number": {
                        "MAX_SAFE_INTEGER": {
                            "__compat": {
                                "support": {
                                    "nodejs": { "version_added": "0.12.0" }
                                }
                            }
                        }
                    },
                    "Temporal": {
                        "__compat": {
                            "support": {
                                "nodejs": { "version_added": "26.0.0" }
                            }
                        }
                    }
                }
            }
        });

        let features = extract_features(&bcd, "nodejs").unwrap();
        let names: Vec<_> = features
            .iter()
            .map(|feature| feature.name.as_str())
            .collect();

        assert!(names.contains(&"fetch"));
        assert!(names.contains(&"Array.toSorted"));
        assert!(names.contains(&"Number.MAX_SAFE_INTEGER"));
        assert!(names.contains(&"Temporal"));
        assert!(!names.contains(&"Array.options_parameter"));
        assert!(!names.contains(&"Array.flagged"));

        let to_sorted = features
            .iter()
            .find(|feature| feature.name == "Array.toSorted")
            .unwrap();
        assert_eq!(to_sorted.version.to_string(), "20.0.0");
        assert!(
            to_sorted
                .detect
                .contains(&GeneratedDetectRule::Property("toSorted".to_string()))
        );
    }
}

File: scripts/generate_mdn_data.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import sys
from dataclasses import dataclass
from pathlib import Path
from urllib.request import urlopen


DEFAULT_OUTPUT_DIR = Path("data/generated/mdn")
DEFAULT_RUNTIMES = ["nodejs", "deno", "bun", "safari", "chrome", "firefox"]
SYNTAX_SECTIONS = [
    "classes",
    "functions",
    "grammar",
    "operators",
    "regular_expressions",
    "statements",
]
IGNORED_RUNTIME_FEATURES = {
    "undefined",
}


@dataclass(frozen=True, order=True)
class Version:
    major: int
    minor: int
    patch: int

    @classmethod
    def parse(cls, value: str) -> Version:
        value = value.strip().removeprefix("v")
        if not value or not value[0].isdigit():
            raise ValueError(f"version is not numeric: {value}")

        parts = value.split(".")
        while len(parts) < 3:
            parts.append("0")
        return cls(*(int(part) for part in parts[:3]))

    def __str__(self) -> str:
        return f"{self.major}.{self.minor}.{self.patch}"


@dataclass(frozen=True)
class RuntimeTarget:
    bcd_id: str
    runtime_name: str


def main() -> int:
    args = parse_args()
    bcd_version = args.bcd_version or Path("data/mdn-bcd.version").read_text().strip()
    bcd = load_bcd(args.input, bcd_version)

    actual = bcd.get("__meta", {}).get("version")
    if actual and actual != bcd_version:
        raise SystemExit(f"expected BCD {bcd_version}, got {actual}")

    args.output_dir.mkdir(parents=True, exist_ok=True)
    for runtime_id in args.runtimes:
        target = runtime_target(runtime_id)
        features = extract_features(bcd, target.bcd_id)
        output = {
            "schema": 1,
            "runtime": target.runtime_name,
            "source": {
                "name": "@mdn/browser-compat-data",
                "version": bcd_version,
                "runtime": target.bcd_id,
            },
            "features": features,
        }
        path = args.output_dir / f"{target.runtime_name}.json"
        path.write_text(json.dumps(output, indent=2, sort_keys=False) + "\n", encoding="utf-8")
        print(
            f"generated {len(features)} features for {target.runtime_name} "
            f"from MDN BCD {bcd_version} -> {path}"
        )

    return 0


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        prog="generate_mdn_data.py",
        description="Generate JSON runtime data artifacts from MDN browser-compat-data.",
    )
    parser.add_argument("--input", type=Path, help="read MDN BCD data.json from disk")
    parser.add_argument(
        "--bcd-version",
        help="MDN @mdn/browser-compat-data version to download",
    )
    parser.add_argument(
        "--output-dir",
        type=Path,
        default=DEFAULT_OUTPUT_DIR,
        help=f"directory where JSON artifacts are written (default: {DEFAULT_OUTPUT_DIR})",
    )
    parser.add_argument(
        "--runtimes",
        default=",".join(DEFAULT_RUNTIMES),
        help="comma-separated BCD runtime ids: nodejs,deno,bun,safari,chrome,firefox",
    )
    args = parser.parse_args()
    args.runtimes = [runtime.strip() for runtime in args.runtimes.split(",") if runtime.strip()]
    return args


def load_bcd(input_path: Path | None, version: str) -> dict:
    if input_path:
        return json.loads(input_path.read_text(encoding="utf-8"))

    url = f"https://unpkg.com/@mdn/browser-compat-data@{version}/data.json"
    with urlopen(url) as response:
        return json.loads(response.read().decode("utf-8"))


def runtime_target(runtime_id: str) -> RuntimeTarget:
    match runtime_id:
        case "nodejs" | "node":
            return RuntimeTarget("nodejs", "node")
        case "deno":
            return RuntimeTarget("deno", "deno")
        case "bun":
            return RuntimeTarget("bun", "bun")
        case "safari":
            return RuntimeTarget("safari", "safari")
        case "chrome":
            return RuntimeTarget("chrome", "chrome")
        case "firefox":
            return RuntimeTarget("firefox", "firefox")
        case _:
            raise SystemExit(f"unsupported BCD runtime id `{runtime_id}`")


def extract_features(bcd: dict, runtime_id: str) -> list[dict]:
    features: dict[str, dict] = {}

    api = bcd.get("api")
    if isinstance(api, dict):
        walk_compat_tree(api, [], "api", runtime_id, features)

    builtins = bcd.get("javascript", {}).get("builtins")
    if isinstance(builtins, dict):
        walk_compat_tree(builtins, [], "builtin", runtime_id, features)

    javascript = bcd.get("javascript", {})
    for section in SYNTAX_SECTIONS:
        syntax = javascript.get(section)
        if isinstance(syntax, dict):
            walk_compat_tree(syntax, [], f"syntax:{section}", runtime_id, features)

    return [features[name] for name in sorted(features)]


def walk_compat_tree(
    value: dict,
    path: list[str],
    source: str,
    runtime_id: str,
    features: dict[str, dict],
) -> None:
    compat = value.get("__compat")
    if isinstance(compat, dict):
        version = support_version(compat, runtime_id)
        name = feature_name(source, path)
        if version and name and is_supported_path(source, path):
            detect = detect_rules(source, path)
            if detect:
                upsert_feature(
                    features,
                    {
                        "name": name,
                        "version": str(version),
                        "detect": detect,
                    },
                )

    for key, child in value.items():
        if key == "__compat" or not isinstance(child, dict):
            continue
        path.append(key)
        walk_compat_tree(child, path, source, runtime_id, features)
        path.pop()


def upsert_feature(features: dict[str, dict], feature: dict) -> None:
    existing = features.get(feature["name"])
    if existing and Version.parse(existing["version"]) <= Version.parse(feature["version"]):
        return
    features[feature["name"]] = feature


def support_version(compat: dict, runtime_id: str) -> Version | None:
    support = compat.get("support", {}).get(runtime_id)
    if support is None:
        return None

    statements = support if isinstance(support, list) else [support]
    versions = []
    for statement in statements:
        if not isinstance(statement, dict):
            continue
        if statement.get("flags"):
            continue
        if statement.get("version_removed") is not None:
            continue
        version_added = statement.get("version_added")
        if not isinstance(version_added, str):
            continue
        try:
            versions.append(Version.parse(version_added))
        except ValueError:
            continue
    return min(versions) if versions else None


def feature_name(source: str, path: list[str]) -> str:
    name = ".".join(path)
    if source.startswith("syntax:"):
        section = source.split(":", 1)[1]
        return f"syntax.{section}.{name}" if name else ""
    return name


def is_supported_path(source: str, path: list[str]) -> bool:
    if source.startswith("syntax:"):
        return bool(path) and all(is_syntax_segment(segment) for segment in path)
    return is_runtime_surface_path(path)


def is_runtime_surface_path(path: list[str]) -> bool:
    if not path:
        return False
    if ".".join(path) in IGNORED_RUNTIME_FEATURES:
        return False
    if len(path) > 1 and path[0] == path[-1]:
        return False
    return all(is_runtime_surface_segment(segment) for segment in path)


def is_runtime_surface_segment(segment: str) -> bool:
    if not segment or segment.startswith("@@") or "-" in segment:
        return False
    if "_" in segment:
        return all(ch == "_" or ch.isdigit() or ("A" <= ch <= "Z") for ch in segment)
    return all(ch == "$" or ch.isalnum() for ch in segment)


def is_syntax_segment(segment: str) -> bool:
    return bool(segment) and all(ch == "_" or ch.isalnum() for ch in segment)


def detect_rules(source: str, path: list[str]) -> list[dict]:
    if not path:
        return []

    name = ".".join(path)
    rules = set()
    if source == "api" or source == "builtin":
        if len(path) == 1:
            rules.add(("Global", name))
        rules.add(("MemberChain", name))
        if source == "builtin" and len(path) >= 2 and is_detectable_property(path[-1]):
            rules.add(("Property", path[-1]))
    elif source.startswith("syntax:"):
        section = source.split(":", 1)[1]
        rules.add(("Syntax", f"{section}.{name}"))

    return [{"kind": kind, "value": value} for kind, value in sorted(rules)]


def is_detectable_property(property_name: str) -> bool:
    return (
        not property_name.startswith("@@")
        and not property_name.startswith("__")
        and bool(property_name)
        and (property_name[0] == "_" or property_name[0].isalpha())
    )


if __name__ == "__main__":
    sys.exit(main())
